cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go

```diff
@@ -313,14 +322,66 @@ func (oi *OVFImporter) importDiskFilesToImages() error {
 		return err
 	}
 
+	oi.params.OsID = osIDValue
+	oi.params.Deadline = oi.params.Deadline.Add(-1 * instanceConstructionTime)
+
 	oi.machineTypeString, err = oi.getMachineType(ovfDescriptor, *oi.params.Project, oi.params.Zone)
 	if err != nil {
 		return err
 	}
 
 	oi.Logger.User(fmt.Sprintf("Will create instance of `%v` machine type.", oi.machineTypeString))
 
-	return oi.importDisks(osIDValue, &diskInfos)
+	var disksNamesPrefix string
+	if oi.params.IsInstanceImport() {
+		disksNamesPrefix = oi.params.InstanceNames
+	} else {
+		disksNamesPrefix = oi.params.MachineImageName
+	}
+
+	if len(diskInfos) > 1 {
+		var dataDiskURIs []string
+		for _, info := range diskInfos[1:] {
+			dataDiskURIs = append(dataDiskURIs, info.FilePath)
+		}
+		disks, err := oi.multiDiskImporter.Import(oi.ctx, oi.params, dataDiskURIs)
+		if err != nil {
+			return err
+		}
+		oi.disks = disks
+	}
+
+	if len(diskInfos) > 0 {
+		if err = oi.importBootDiskImage(diskInfos[0], disksNamesPrefix); err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+func (oi *OVFImporter) importBootDiskImage(bootDiskInfos ovfutils.DiskInfo, disksNamesPrefix string) error {
+
+	oi.Logger.User(fmt.Sprint("Importing boot Disk Image ..."))
+
+	imageName := fmt.Sprintf("%s-%s", disksNamesPrefix, "boot-image")
+	request, err := oi.buildBootDiskImageImportRequest(imageName, bootDiskInfos.FilePath)
+
+	if oi.imageImporter == nil {
+		oi.imageImporter, err = importer.NewImporter(request, oi.computeClient, oi.storageClient, oi.Logger)
```

Comment:
Could you clarify why do we need here a lazy assignment? As I see this method is called only once. Have you checked a way to avoid having a nullable field here?

Reply:
This part is slightly not clear to me, the method `importBootDiskImage` accepts two arguments: `bootDiskInfos` and `disksNamesPrefix`.

- `disksNamesPrefix` is quite a straightforward argument which value we can define in the early beginning;
- `bootDiskInfos` is constructed **before** importing disks, and uses only `ovfutils.GetOVFDescriptorAndDiskPaths`. `ofvDescriptorLoader` is created in the importer constructor: https://github.com/GoogleCloudPlatform/compute-image-tools/blob/1675a8ccca681f85284255749386ec6e8d786567/cli_tools/gce_ovf_import/ovf_importer/ovf_importer.go#L123 It looks to me that we can move the building of `imageImporter` there, do I miss anything?

Reply:
I'm talking about **NewImporter** function not **importBootDiskImage**, NewImporter accepts the ImageImportRequest object as a parameter, and I've to create the dataDisks before building the boot-disk ImageImportRequest as it uses them.

Reply:
Well, in this particular context it means exactly the same:

```go
func (oi *OVFImporter) importBootDiskImage(bootDiskInfos ovfutils.DiskInfo, disksNamesPrefix string) error {

	oi.Logger.User(fmt.Sprint("Importing boot Disk Image ..."))

	imageName := fmt.Sprintf("%s-%s", disksNamesPrefix, "boot-image")
	request, err := oi.buildBootDiskImageImportRequest(imageName, bootDiskInfos.FilePath)

	if oi.imageImporter == nil {
		oi.imageImporter, err = importer.NewImporter(request, oi.computeClient, oi.storageClient, oi.Logger)
		if err != nil {
...
```

To create an image importer you need the following set of arguments: `request`, `oi.computeClient`, `oi.storageClient`, `oi.Logger`.

`oi.computeClient`, `oi.storageClient`, `oi.Logger` are created in the `NewOVFImporter` method, therefore the only argument you need to consider is `request`.

`request` is created by a call to `oi.buildBootDiskImageImportRequest(imageName, bootDiskInfos.FilePath)` that requires `imageName` and `bootDiskInfos.Filepath`.
 - `imageName` is created from `disksNamesPrefix` that is passed to `importBootDiskImage`;
 - `bootDiskInfos` is passed to `importBootDiskImage` as well.

And as I see from the code, you don't have to create data disks to initialise `disksNamesPrefix` and `bootDiskInfos`:

```go
	var disksNamesPrefix string
	if oi.params.IsInstanceImport() {
		disksNamesPrefix = oi.params.InstanceNames
	} else {
		disksNamesPrefix = oi.params.MachineImageName
	}
```

```go
	ovfDescriptor, diskInfos, err := ovfutils.GetOVFDescriptorAndDiskPaths(
		oi.ovfDescriptorLoader, ovfGcsPath)
	if err != nil {
		return err
	}
```

Therefore it looks to me that the importer can be created inside the `NewOVFImporter` method. Do I miss anything?

Reply:
But DataDisks which is part of the request is constructed after calling this function "oi.multiDiskImporter.Import()"

Reply:
more details, **importBootDiskImage** function call **oi.buildBootDiskImageImportRequest** function to construct the request, it requires creating the data disks first becasue dataDisks are part of the request, it got initialized [here](https://github.com/GoogleCloudPlatform/compute-image-tools/pull/1864/files#diff-1ef6e3baf6905b5deba922e96bde4ae98152db2bce8590088b7d3e33add4695eR547)